### ISSUE TITLE
Retry.get_retry_after does't respect time-out

### ISSUE THREAD

**Issue by krukas** [shazow/urllib3#1338]
Maybe a weird case but if a website gives back an Retry-After that is higher then given timeout it will accept that. 

In my case some website gives back a retry of 3600 and so my checkup tasks was waiting for 1 hour to try it one more time.

Maybe ignore large values or at least if Retry-After is bigger then timeout then use timeout or ignore Retry-After

**Comment by SethMichaelLarson** [comment]
Could you recreate a testcase using httpbin so we can verify this? :)

**Comment by sigmavirus24** [comment]
> Maybe ignore large values or at least if Retry-After is bigger then timeout 

`timeout` is not a wall-clock timeout. `timeout` applies to time it takes to connect to the server and time it takes to read the contents.

The logic you're asking for could instead be implemented in a subclass of `Retry`, if I remember correctly. It's not inherently correct when you consider what `timeout` does for the library.

**Comment by haikuginger** [comment]
I agree with @sigmavirus24. I would consider a pull request that adds an additional kwarg (something like `max_retry_wait_length`) that raises an exception (with sufficient information to allow retrying the request after the server-given interval) when the response wants to wait longer than a specified interval.

**Comment by Furox-Art** [comment]
I've opened #5220 implementing the opt-in behavior discussed above: a `raise_on_retry_after_max` flag on `Retry` that raises a new `RetryAfterMaxExceededError` (carrying `retry_after` and `max_wait`) instead of silently capping, keeping the default behavior unchanged. Happy to adjust the name/shape if maintainers prefer a different API (e.g. `max_retry_wait_length` as mentioned earlier in the thread).

**Comment by WUZICANGJIE** [comment]
AI-assisted validation note after preparing a local prototype and comparing #5220: the main opt-in behavior is already covered there. One concrete hardening point is exception serialization. Reproducing its current two-argument exception constructor in isolation, `pickle.loads(pickle.dumps(RetryAfterMaxExceededError(3600, 60)))` raises `TypeError: ... missing 1 required positional argument: 'max_wait'`. A `__reduce__` returning `(type(self), (self.retry_after, self.max_wait))`, with a round-trip test, preserves the scheduling information.

I also ran loopback HTTP tests on my independent prototype for pool-level 503 retry, pool-level 302 redirect, and PoolManager 503 retry: with `preload_content=False` and a single blocking connection, an excessive Retry-After raises without sleeping or sending a second request, and a subsequent request succeeds (checking that the response was drained/released). The retry/exception tests plus these additions passed locally: 73 passed, 18 Windows timezone skips. This is not a test result for #5220 itself.

Would a focused exception-hardening and transport-test contribution alongside the existing implementation be useful and eligible for a portion of this issue's bounty? I have not opened another competing PR or claimed payment. If the existing contributors already cover this, I will leave the work with them.
